src/main/java/org/haffson/adventofcode/days/day02/Day02.java

```diff
@@ -0,0 +1,177 @@
+            this.pwd = pwd;
+        }
+    }
+
+    public List<Data> getData(List<String> rawData) {
+
+        List<Data> dataArrayList = new ArrayList<>();
+
+        for (int i = 0; i < rawData.size(); i++) {
+
+            String line = rawData.get(i);
+
+            // match patterns to determine min and max number of the letter in pwd, the searched letter and pwd itself
+            String pattern = "(\\d+)-(\\d+)\\s+(\\w):\\s+(\\w*)";
+            int min = Integer.parseInt(line.replaceAll(pattern, "$1"));
+            int max = Integer.parseInt(line.replaceAll(pattern, "$2"));
+            String letter = line.replaceAll(pattern, "$3");
+            String pwd = line.replaceAll(pattern, "$4");
+
+            dataArrayList.add(new Data(min, max, letter, pwd));
+        }
+        return dataArrayList;
+    }
+
+
+    /**
+     * Primary method for Day 1, Part 1.
+     * Gets the number of correct passwords from a list
+     *
+     * @return the number of correct pwds in list
+     */
+
+    private int getNumberPwd1() throws FileNotFoundException {
+        // get data
+        List<Data> data = getData(rawData);
+
+        int countCorrPwd = 0; // answer to puzzle day 2.1: number of correct passwords in list
+
+        // loop through passwords
+        for (int i=0; i < data.size(); i++ ) {
+            int count = 0;  // number of letter appearance in pwd
+
+            // count how often specific letter appears in pwd
+            for (int j = 0; j < data.get(i).pwd.length(); j++) {
+                if (data.get(i).pwd.charAt(j) == data.get(i).letter.charAt(0)) count++;
+            }
+             // check if count meets criteria for correct pwd
+            if (count >= data.get(i).min && count <= data.get(i).max) {
+                countCorrPwd++;
+            }
+        }
+        return countCorrPwd;
+    }
+
+    // if raw data is not read via file, e.g. test data
+    public void addInput(List<String> rawData) {
+        this.rawData = rawData;
+    }
+
+    private int getNumberPwd2() throws FileNotFoundException {
+        // get data
+        List<Data> data = getData(rawData);
+
+        int countCorrPwd = 0; // answer to puzzle day 2.2: number of correct passwords in list
+
+        // loop through passwords
+        for (int i=0; i < data.size(); i++ ) {
+
+            boolean onlyFirstPosition = data.get(i).pwd.charAt(data.get(i).min - 1) == data.get(i).letter.charAt(0) && data.get(i).pwd.charAt(data.get(i).max - 1) != data.get(i).letter.charAt(0);
+            boolean onlySecondPosition = data.get(i).pwd.charAt(data.get(i).max - 1) == data.get(i).letter.charAt(0) && data.get(i).pwd.charAt(data.get(i).min - 1) != data.get(i).letter.charAt(0);
+
+            // letter must appear only once (either only(!) at first or only(!) at second position)
+            if ((onlyFirstPosition && !onlySecondPosition) || (!onlyFirstPosition && onlySecondPosition)) {
+                countCorrPwd++;
+            }
+        }
```

Comment:
The same simplification is also possible here:
```suggestion
        // loop through passwords
        for (Data datum : data) {

            boolean onlyFirstPosition = datum.pwd.charAt(datum.min - 1) == datum.letter.charAt(0) && datum.pwd.charAt(datum.max - 1) != datum.letter.charAt(0);
            boolean onlySecondPosition = datum.pwd.charAt(datum.max - 1) == datum.letter.charAt(0) && datum.pwd.charAt(datum.min - 1) != datum.letter.charAt(0);

            // letter must appear only once (either only(!) at first or only(!) at second position)
            if ((onlyFirstPosition && !onlySecondPosition) || (!onlyFirstPosition && onlySecondPosition)) {
                countCorrPwd++;
            }
        }
```

Maybe it is not so simple at first to understand why this is supposed to be simpler, because some magic is hidden behind the enhanced for, but it gets a little more readable because it is just shorter to read.

One thing about code is that it is read way more often than it is written. This is why developers have such a strong opinion on formatting, style, and naming of variables.